### PR TITLE
chore: change to vitest and remove cjs bundling.

### LINKED ISSUE #1055: evaluate whether we still need commonjs support
Since we originally stood up this library the ecosystem has moved a lot:
- nodejs supports ESM, and all supported versions have them enabled by default now
- major community projects have moved to ESM only, including some in our build infrastructure (see #1041 where we can't upgrade to chai 5 because we still have commonjs based tests)
- we've established that the sub modules we have (fluent API, sub-namespaces for models...) can only be consumed in ESM (when the generated client is published as its own package) when working on msgraph-sdk-typescript.
- we're still on the 2.X branch of node-fetch because 3.X [moved to ESM](https://github.com/node-fetch/node-fetch/blob/main/docs/v3-UPGRADE-GUIDE.md)

IMHO we should remove commonjs support after testing our main scenarios (SPFx, TeamsFX....) and BEFORE the GA so people don't end up in confusing variants + configuration misalignment issues.

### LINKED ISSUE #1072: Polymorphism of buffer usage
The libraries here make use of Buffer.from multiple places
https://github.com/microsoft/kiota-typescript/blob/dfaf4045855f358a44f1fa91b5b85da901b5d24e/packages/authentication/azure/src/azureIdentityAccessTokenProvider.ts#L85
https://github.com/microsoft/kiota-typescript/blob/dfaf4045855f358a44f1fa91b5b85da901b5d24e/packages/serialization/json/src/jsonParseNode.ts#L59
https://github.com/microsoft/kiota-typescript/blob/dfaf4045855f358a44f1fa91b5b85da901b5d24e/packages/serialization/json/src/jsonSerializationWriter.ts#L19
https://github.com/microsoft/kiota-typescript/blob/dfaf4045855f358a44f1fa91b5b85da901b5d24e/packages/serialization/text/src/textParseNode.ts#L31
https://github.com/microsoft/kiota-typescript/blob/dfaf4045855f358a44f1fa91b5b85da901b5d24e/packages/serialization/text/src/textSerializationWriter.ts#L20

This API is NOT AVAILABLE on browsers, we're using it to perform base64 encoding/decoding.

We should instead:
- define two base64 encode/decode methods in abstractions
- in the implementations test whether Buffer.from is available, if so, use that.
- in the implementations test whether atob/btoa are available, if so, use that.
- if none of those APIs are available, throw
- replace the calls to Buffer.from by the corresponding method.

### PR DESCRIPTION
**Pull request by musale**
closes #1072, closes #1055

This is a change that is part of a bigger goal to improve our confidence in browser testing and update the codebase to adapt to node/browser environment without throwing errors:
- adds Vitest as the test framework that has browser testing.
- drops karma. eventually drop add-ins like Sinon used for mocking and completely build on Vitest.
- updates test imports to use Vitest API.
- adds browser testing for the packages.
- adds utility function to check the environment(node/browser) the code is running in.
- Fixes browser incompatible code in the codebase.
- Removes common js bundling tooling and support.

### Status

- [x] `abstractions` package is updated to vitest and has browser tests
- [x] `authentication/azure` package is updated to vitest and has a trigger for browser tests that pass
- [x] `authentication/spfx` is updated
- [x] `http` is updated.
- [x] `serialization/*` is updated.

### CODE REVIEW

File: packages/abstractions/test/common/authentication/validateProtocolTest.ts
Comment (baywet) [inline packages/abstractions/test/common/authentication/validateProtocolTest.ts:5]: adding this comment as a reminder

**Review by baywet — APPROVED** [review]
great work so far! can't wait for this to be finalized

**Comment by musale** [comment]
@baywet I'm having a blocker with completing this. When running the steps to build and test, there's an error with building that comes from having build instructions in tsconfig that outputs commonjs. I think it would be a good idea to make a decision on #1055 to unblock this. What do you think?

**Comment by baywet** [comment]
Agreed, let's get confirmation on the PM side.
If Seb doesn't reply within a few days, please reach out to him directly

File: packages/abstractions/src/apiError.ts
Comment (baywet) [inline packages/abstractions/src/apiError.ts:1]: for my education: are doc comment ignored for interfaces?

Comment (musale) [reply packages/abstractions/src/apiError.ts:1]: I don't understand the context of your question @baywet

Comment (baywet) [reply packages/abstractions/src/apiError.ts:1]: I saw that you removed the doc comment, can you explain why?

File: .github/workflows/build_test_validate.yml
Comment (baywet) [inline .github/workflows/build_test_validate.yml:13]: revert

File: packages/abstractions/src/apiError.ts
Comment (musale) [reply packages/abstractions/src/apiError.ts:1]: Oh, I see! I didn't remove that intentionally; it must've slipped out during one of the merges. I'll revert it too.

**Comment by koros** [comment]
Hi @musale,
Can you please give instructions on how to run the build and tests locally. this is what I get:

### `npm run build`
npm-run-build

### `npm run test`
npm-run-test

**Comment by musale** [comment]
> Hi @musale, Can you please give instructions on how to run the build and tests locally. this is what I get:
> 
> ### `npm run build`
> npm-run-build
> 
> ### `npm run test`
> npm-run-test

Hi @koros this CI/CD pipeline and build is adapted for a Linux environment. I see you're using PowerShell and this will fail for commands like `clean` and path resolutions. Please use WSL or a Linux environment to test this.

File: packages/authentication/azure/package.json
Comment (baywet) [inline packages/authentication/azure/package.json:34]: what's the rationale behind switching to a star reference?

File: packages/abstractions/src/store/backingStore.ts
Comment (baywet) [inline packages/abstractions/src/store/backingStore.ts:4]: still missing the doc comment here (and a couple of other places)

**Review by baywet — CHANGES_REQUESTED** [review]
Getting there, still a couple of remarks

File: packages/authentication/azure/package.json
Comment (musale) [reply packages/authentication/azure/package.json:34]: Since we are using Lerna to manage the monorepo, in local development using the * allows using the latest changes in a package.

Comment (musale) [reply packages/authentication/azure/package.json:34]: Yes, the published version will use the latest version of abstractions. Is that going to be problematic?

Comment (baywet) [reply packages/authentication/azure/package.json:34]: no, that's what I expected. Thanks for confirming. Resolving

**Review by baywet — APPROVED** [review]
@musale there are new conflicts to resolve, but once that is done, I think we're good to merge here. Thanks for the great work!

File: packages/serialization/json/src/jsonParseNode.ts
Comment (baywet) [inline packages/serialization/json/src/jsonParseNode.ts:55]: probably a regression during the merge here

Comment (musale) [reply packages/serialization/json/src/jsonParseNode.ts:55]: Ah yeah, good catch!

**Comment by koros** [comment]
> > Hi @musale, Can you please give instructions on how to run the build and tests locally. this is what I get:
> > ### `npm run build`
> > npm-run-build
> > ### `npm run test`
> > npm-run-test
> 
> Hi @koros this CI/CD pipeline and build is adapted for a Linux environment. I see you're using PowerShell and this will fail for commands like `clean` and `tsconfig` file path resolutions. Please use WSL or a Linux environment to test this.

@musale this seems to resolve the issue on Win OS
``` JSON
    "build:@microsoft/*": "lerna run build --scope \"@microsoft/*\"",
    "clean": "lerna run --parallel --stream --scope \"@microsoft/*\" clean",
    "test:browser": "lerna run --scope \"@microsoft/*\" test:browser",
    "test:node": "lerna run --scope \"@microsoft/*\" test:node",
```

File: packages/test/package.json
Comment (koros) [inline packages/test/package.json:17]:
```suggestion
    "@microsoft/kiota-serialization-text": "*",
    "@microsoft/kiota-serialization-multipart": "*"
```

File: packages/test/package.json
Comment (koros) [reply packages/test/package.json:8]: We should also probably change the `name` to `@microsoft/test-kiota-typescript-libraries`.
at the moment the whole package gets skipped when running build and tests because name doesn't match the filter criteria `"@microsoft/*"`

Comment (musale) [reply packages/test/package.json:8]: @koros the `test` package is not a Microsoft namespace package, it's only used to run integrated tests. Running it as part of the Lerna workspace commands also introduces unnecessary commands like `build` and `clean`. It also has dependency issues needing IDs and secrets which, during Lerna runtime, we'll have to pass them on the test commands and it's not a recommended practice. As it is, running it as its own step in the CI/CD pipeline is sufficient.

**Review by baywet — APPROVED** [review]

**Review by koros — APPROVED** [review]

**Pull request merged**
